rclcpp/src/rclcpp/time_source.cpp

```diff
@@ -42,15 +43,79 @@ TimeSource::TimeSource()
 {
 }
 
-void TimeSource::attachNode(std::shared_ptr<rclcpp::node::Node> node)
+void TimeSource::attachNode(rclcpp::node::Node::SharedPtr node)
 {
-  node_ = node;
+  attachNode(
+    node->get_node_base_interface(),
+    node->get_node_topics_interface(),
+    node->get_node_graph_interface(),
+    node->get_node_services_interface());
+}
+
+void TimeSource::attachNode(
+  const rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base_interface,
+  const rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr node_topics_interface,
+  const rclcpp::node_interfaces::NodeGraphInterface::SharedPtr node_graph_interface,
+  const rclcpp::node_interfaces::NodeServicesInterface::SharedPtr node_services_interface)
+{
+  node_base_ = node_base_interface;
+  node_topics_ = node_topics_interface;
+  node_graph_ = node_graph_interface;
+  node_services_ = node_services_interface;
   // TODO(tfoote): Update QOS
-  clock_subscription_ = node_->create_subscription<builtin_interfaces::msg::Time>(
-    "clock", std::bind(&TimeSource::clock_cb, this, std::placeholders::_1),
-    rmw_qos_profile_default);
 
-  parameter_client_ = std::make_shared<rclcpp::parameter_client::AsyncParametersClient>(node);
+  const std::string & topic_name = std::string("clock");
+
+  rclcpp::callback_group::CallbackGroup::SharedPtr group;
+  using rclcpp::message_memory_strategy::MessageMemoryStrategy;
+  auto msg_mem_strat = MessageMemoryStrategy<MessageT, Alloc>::create_default();
+  auto allocator = std::make_shared<Alloc>();
+
+  // using MessageT_arg = builtin_interfaces::msg::Time::SharedPtr;
+  // using Alloc_arg = std::allocator<void>;
+  // using SubscriptionT_arg = rclcpp::subscription::Subscription<MessageT, Alloc_arg>;
+
+  // std::function<void(const std::shared_ptr<MessageT>)> cb =
+  //   std::bind(&TimeSource::clock_cb, this, std::placeholders::_1);
+    
+  auto cb = std::bind(&TimeSource::clock_cb, this, std::placeholders::_1);
+  rclcpp::any_subscription_callback::AnySubscriptionCallback<MessageT, Alloc> any_cb(allocator);
+  any_cb.set(std::move(cb));
+
+  clock_subscription_ = rclcpp::create_subscription < MessageT, CallbackT, Alloc, SubscriptionT>
```

Comment:
@Karsten1987 @mikaelarguedas This is the method we were looking at and I still cannot get to compile.

@wjwwood Is this the approach you'd recommend here? In the comments you can see various other approaches that I've looked at as well.  It was working find using the full node api [previously](https://github.com/ros2/rclcpp/pull/407/files#diff-f5fd9d522b0a1738cae064711fe7ce43L49)

Reply:
You don't need to deal with the AnySubscriptionCallback at this point here. The actual factory does take care of it later:
https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/subscription_factory.hpp#L78-L80

That being said, you have to deduce the template argument for `CallbackT` manually. In this case, you could either simply use `decltype` and let c++11 take care of it or specify it manually as `std::function<void(const std::shared_ptr<MessageT>)>`

So the following code compiles, haven't tested it if it actually behaves correctly :)

```
auto cb = std::bind(&TimeSource::clock_cb, this, std::placeholders::_1);

//clock_subscription_ = rclcpp::create_subscription < MessageT, std::function<void(const std::shared_ptr<MessageT>)>, Alloc, SubscriptionT>
    
clock_subscription_ =
  rclcpp::create_subscription< MessageT, decltype(cb), Alloc, SubscriptionT>
(
  node_topics_.get(),
  topic_name,
  std::move(cb),
  rmw_qos_profile_default,
  group,
  false,
  false,
  msg_mem_strat,
  allocator);
```

Reply:
Thanks @Karsten1987 with that change it's passing the unit tests again.